### PR TITLE
Scatter/gather for pointers

### PR DESCRIPTION
**Pull request by calebzulawski**
Also provides a solution for #205

### CODE REVIEW

**Review by programmerjake — APPROVED** [review]

**Comment by programmerjake** [comment]
> Also provides a solution for #286

not really, you still can't do:
`v.gather_or(mask, Simd::<MaybeUninit<u32>, 4>::uninit())`
or:
```rust
#[repr(C, align(16))]
struct S {
    x: f32,
    y: f32,
    z: f32,
}

fn simd_add(a: S, b: S) -> S {
    // this is safe
    let a: Simd<MaybeUninit<f32>, 4> = unsafe { mem::transmute(a) };
    let b: Simd<MaybeUninit<f32>, 4> = unsafe { mem::transmute(b) };
    let v: Simd<MaybeUninit<f32>, 4> = a + b;
    unsafe { mem::transmute(v) }
}
```

**Comment by calebzulawski** [comment]
Whoops, I referenced the wrong `MaybeUninit` issue.  I meant #205

File: crates/core_simd/src/vector.rs [hunk 1]
```diff
@@ -364,8 +364,71 @@ where
         let base_ptr = Simd::<*const T, LANES>::splat(slice.as_ptr());
         // Ferris forgive me, I have done pointer arithmetic here.
         let ptrs = base_ptr.wrapping_add(idxs);
-        // Safety: The ptrs have been bounds-masked to prevent memory-unsafe reads insha'allah
-        unsafe { intrinsics::simd_gather(or, ptrs, enable.to_int()) }
+        // Safety: The caller is responsible for determining the indices are okay to read
+        unsafe { Self::gather_select_ptr(ptrs, enable, or) }
+    }
+
+    /// Read pointers elementwise into a SIMD vector vector.
```
Comment (workingjubilee) [inline crates/core_simd/src/vector.rs:371]: "vector vector"?

File: crates/core_simd/src/vector.rs [hunk 2]
```diff
@@ -473,10 +536,63 @@ where
...
+    /// assert_eq!(values, [7, 5, 3, 6]);
+    /// ```
+    #[inline]
+    #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    pub unsafe fn scatter_ptr(self, dest: Simd<*mut T, LANES>) {
+        // Safety: The caller is responsible for upholding all invariants
+        unsafe { self.scatter_select_ptr(dest, Mask::splat(true)) }
+    }
+
+    /// Conditionally write pointers elementwise into a SIMD vector vector.
```
Comment (workingjubilee) [inline crates/core_simd/src/vector.rs:569]: "vector vector"?

File: crates/core_simd/src/vector.rs [hunk 3]
```diff
@@ -473,10 +536,63 @@ where
...
+    /// The mask `enable`s all `true` lanes and disables all `false` lanes.
+    /// If a lane is disabled, the writing that lane is skipped.
```
Comment (workingjubilee) [inline crates/core_simd/src/vector.rs:571]:
```suggestion
    /// If a lane is disabled, writing to that lane is skipped.
```
or
```suggestion
    /// If a lane is disabled, the write to that lane is skipped.
```

**Pull request merged**